File: examples/python/e2e_live.py
"""Live end-to-end check against the X Chat API.

Uses the example's real ChatCore (chat_xdk binding) and XChatClient (XDK).
Run manually after `pip install -r requirements.txt`:

    CHATXDK_E2E=1 X_ACCESS_TOKEN=... CHAT_PRIVATE_KEYS_B64=... \\
    CHAT_SIGNING_KEY_VERSION=... CHAT_CONVERSATION_ID=... python e2e_live.py

Flow (each numbered step asserts against the live API):
  1. batch-decrypt inbound history (pagination when a second page exists)
  2. rotate the conversation key (prepare -> POST /keys -> decrypt own CKCE)
  3. send a threaded reply with an entity + TTL under the rotated key,
     fetch it back, decrypt it via the single-event path, and verify it
  4. react to the sent message (add + remove), decrypting the add back

Optional extras:
  CHATXDK_E2E_MEDIA=1   also stream-encrypts a media blob, uploads it,
                        sends a message referencing it, then downloads and
                        stream-decrypts it back to the original bytes
  CHATXDK_E2E_GROUPS=1  also creates a group (two-signature create), sends a
                        group message, and adds the 1:1 partner as a member
  CHAT_PIN=...          also unlocks keys via live Juicebox and checks they
                        match the blob-loaded ones
"""

from __future__ import annotations

import os
import time

from chat_core import ChatCore, message_text, prep_to_request
from x_api import XChatClient


def _signing_from(pk: dict, user_id: str) -> dict:
    return {
        "user_id": user_id,
        "public_key_version": str(pk.get("public_key_version") or ""),
        "public_key": pk.get("signing_public_key") or "",
        "identity_public_key": pk.get("public_key") or "",
        "identity_public_key_signature": pk.get("identity_public_key_signature") or "",
    }


def _key_entries(pks: list[dict], user_id: str) -> list[dict]:
    """Public-keys response -> the flat entries the prepare methods take."""
    return [
        {
            "user_id": user_id,
            "public_key": pk.get("public_key") or "",
            "key_version": str(pk.get("public_key_version") or ""),
        }
        for pk in pks
    ]


def _await_decrypted(
    api: XChatClient,
    core: ChatCore,
    conversation_id: str,
    conv_keys: dict[str, bytes],
    signing: list[dict],
    message_id: str,
    tries: int = 10,
) -> dict:
    """Poll the conversation until the event for ``message_id`` lands, and
    return it decrypted via the single-event path (``decrypt_one``).

    The target envelope is matched by its raw event id before decrypting, so a
    decrypt failure on our own event (e.g. a broken sign->verify loop) surfaces
    in the timeout message instead of being silently swallowed."""
    last_err: Exception | None = None
    for _ in range(tries):
        page = api.get_events(conversation_id, max_results=25)
        for e in page.get("data") or []:
            if not e.get("encoded_event"):
                continue
            is_target = str(e.get("id") or "") == message_id
            try:
                one = core.decrypt_one(e["encoded_event"], conv_keys, signing)
            except Exception as err:
                if is_target:
                    last_err = err
                continue
            if is_target or str(one.get("id") or "") == message_id:
                if not one.get("sequence_id"):
                    # The REST item's id IS the event's sequence id.
                    one["sequence_id"] = e.get("id")
                return one
        time.sleep(1)
    raise AssertionError(
        f"event for sent message {message_id!r} never appeared"
        + (f" (last decrypt error: {last_err})" if last_err else "")
    )


def main() -> None:
    token = os.environ["X_ACCESS_TOKEN"]
    blob = os.environ["CHAT_PRIVATE_KEYS_B64"]
    version = os.environ.get("CHAT_SIGNING_KEY_VERSION", "1")
    conversation_id = os.environ["CHAT_CONVERSATION_ID"]

    core = ChatCore()
    core.load_keys(blob, version)
    api = XChatClient(token)
    my_id = api.get_my_user_id()
    # All signed actions below resolve their sender from the session identity.
    core.set_identity(my_id)

    # -- 1. Inbound history: batch decrypt (+ pagination when available) ----
    page = api.get_events(conversation_id, max_results=10)
    raw = list(page.get("data") or [])
    next_token = (page.get("meta") or {}).get("next_token")
    if next_token:
        page2 = api.get_events(conversation_id, max_results=10, pagination_token=next_token)
        raw2 = list(page2.get("data") or [])
        ids1 = {str(e.get("id")) for e in raw}
        assert raw2 and not ids1 & {str(e.get("id")) for e in raw2}, "pagination made no progress"
        raw += raw2
        print(f"pagination: fetched second page with {len(raw2)} events")

    ids = {my_id} | {str(e.get("sender_id")) for e in raw if e.get("sender_id")}
    signing: list[dict] = []
    pks_by_user: dict[str, list[dict]] = {}
    for uid in ids:
        try:
            pks = api.get_public_keys(uid)
            pks_by_user[uid] = pks
            signing.extend(_signing_from(pk, uid) for pk in pks)
        except Exception:
            pass

    events_b64 = [e["encoded_event"] for e in raw if e.get("encoded_event")]
    batch = core.decrypt_batch(events_b64, signing)
    decrypted = sum(1 for m in batch["messages"] if message_text(m["event"]))
    conv_keys = dict(batch["conversation_keys"]["keys"])
    print(f"live inbound messages decrypted: {decrypted}; conversation keys: {len(conv_keys)}")
    assert decrypted > 0, "expected to decrypt at least one live message"

    # Canonical conversation_id + partner id from the decrypted events.
    canonical_conv = conversation_id
    partner_id = next(iter(ids - {my_id}), None)
    last_inbound_seq = None
    for m in batch["messages"]:
        ev = m["event"] or {}
        if ev.get("conversation_id"):
            canonical_conv = ev["conversation_id"]
        if ev.get("type") == "Message" and str(ev.get("sender_id")) != my_id:
            last_inbound_seq = ev.get("sequence_id") or last_inbound_seq
    assert partner_id, "expected a conversation partner among the senders"

    # -- 2. Key rotation: prepare -> POST /keys -> decrypt own CKCE ---------
    both_keys = _key_entries(pks_by_user.get(my_id) or [], my_id) + _key_entries(
        pks_by_user.get(partner_id) or [], partner_id
    )
    prep = core.prepare_conversation_key_change(both_keys)
    resp = api.add_conversation_keys(conversation_id, prep_to_request(prep, core.public_keys()["signing"]))
    data = resp.get("data") or {}
    assert data.get("sequence_id") or data.get("conversation_key_change_sequence_id"), (
        f"key rotation not acknowledged: {resp}"
    )
    print(f"rotated conversation key to version {prep['conversation_key_version']}"
          + (f"; server conversation_id: {data['conversation_id']}" if data.get("conversation_id") else ""))

    # The rotated key becomes the sending key; re-fetch (polling briefly, in
    # case the CKCE has not propagated yet) so our own CKCE decrypts and the
    # cache includes the new version.
    kv = prep["conversation_key_version"]
    for _ in range(5):
        page = api.get_events(conversation_id, max_results=10)
        events_b64 = [e["encoded_event"] for e in (page.get("data") or []) if e.get("encoded_event")]
        batch = core.decrypt_batch(events_b64, signing)
        conv_keys = dict(batch["conversation_keys"]["keys"])
        if kv in conv_keys:
            break
        time.sleep(1.5)
    assert kv in conv_keys, f"own rotated CKCE (version {kv}) did not decrypt+verify"
    key = conv_keys[kv]

    # -- 3. Send under the rotated key; fetch back; single-event decrypt ----
    marker = f"chat-xdk e2e [python] {int(time.time())}"
    body = core.encrypt_reply(
        conversation_id=canonical_conv,
        text=f"@user {marker}",
        conversation_key=key,
        conversation_key_version=kv,
        reply_to_sequence_id=last_inbound_seq,
        entities=[(0, 5, "mention")],
        ttl_msec=24 * 60 * 60 * 1000,
    )
    api.send_message(canonical_conv, body)
    print(f"sent live encrypted message: {marker!r}")

    one = _await_decrypted(api, core, conversation_id, conv_keys, signing, body["message_id"])
    assert message_text(one) == f"@user {marker}", f"round-trip text mismatch: {one}"
    assert one.get("verified") is True, "own sent message failed signature verification"
    seq = str(one.get("sequence_id") or "")
    assert seq, "sent message has no sequence id to react to"
    print("sent message decrypted + verified via the single-event path")

    # -- 4. Reactions: add (round-trip) then remove --------------------------
    add = core.encrypt_reaction(
        add=True,
        conversation_id=canonical_conv,
        target_message_sequence_id=seq,
        emoji="\U0001f44d",
        conversation_key=key,
        conversation_key_version=kv,
    )
    api.send_message(canonical_conv, add)
    one = _await_decrypted(api, core, conversation_id, conv_keys, signing, add["message_id"])
    content = one.get("content") or {}
    assert content.get("content_type") == "Reaction" and content.get("emoji") == "\U0001f44d", (
        f"expected a Reaction event, got {content}"
    )
    assert one.get("verified") is True, "reaction failed signature verification"
    print("reaction add decrypted + verified")

    remove = core.encrypt_reaction(
        add=False,
        conversation_id=canonical_conv,
        target_message_sequence_id=seq,
        emoji="\U0001f44d",
        conversation_key=key,
        conversation_key_version=kv,
    )
    api.send_message(canonical_conv, remove)
    print("reaction remove sent")

    # -- 5. Optional: media — stream-encrypt, upload, send, download, decrypt
    if os.environ.get("CHATXDK_E2E_MEDIA") == "1":
        # A deterministic multi-chunk payload, so the incremental encryptor
        # emits several frames and any corruption is byte-attributable.
        plaintext = bytes((i * 31 + 7) % 256 for i in range(300_000))
        ciphertext = core.encrypt_media(plaintext, key)
        media_hash_key = api.upload_media(canonical_conv, ciphertext)
        print(f"encrypted media uploaded: {media_hash_key} ({len(ciphertext)} bytes)")

        media_msg = core.encrypt_reply(
            conversation_id=canonical_conv,
            text=f"chat-xdk e2e media [python] {int(time.time())}",
            conversation_key=key,
            conversation_key_version=kv,
            attachments=[
                {
                    "attachment_type": "media",
                    "media_hash_key": media_hash_key,
                    "width": 0,
                    "height": 0,
                    "filesize_bytes": len(plaintext),
                    "filename": "e2e.bin",
                    "media_type": 5,
                }
            ],
            ttl_msec=24 * 60 * 60 * 1000,
        )
        api.send_message(canonical_conv, media_msg)
        one = _await_decrypted(api, core, conversation_id, conv_keys, signing, media_msg["message_id"])
        assert one.get("verified") is True, "media message failed signature verification"
        atts = (one.get("content") or {}).get("attachments") or []
        got_key = next(
            (a["media"].get("media_hash_key") for a in atts if a.get("media")), None
        )
        assert got_key == media_hash_key, f"attachment did not round-trip: {atts}"

        downloaded = api.download_media(canonical_conv, media_hash_key)
        assert core.decrypt_media(downloaded, key) == plaintext, (
            "downloaded media did not decrypt to the original bytes"
        )
        print("media downloaded + stream-decrypted to the original bytes")

    # -- 6. Optional: group create + message + member add --------------------
    if os.environ.get("CHATXDK_E2E_GROUPS") == "1":
        _groups_flow(api, core, my_id, partner_id, both_keys, signing)

    # -- 7. Optional: live Juicebox unlock -----------------------------------
    pin = os.environ.get("CHAT_PIN")
    if pin:
        config, latest_version = api.get_juicebox_config(my_id)
        jb = ChatCore()
        jb.unlock(config, pin, latest_version)
        assert jb.public_keys()["identity"] == core.public_keys()["identity"], (
            "Juicebox-unlocked identity key differs from the blob-loaded one"
        )
        print("juicebox live unlock: keys match")

    print("E2E PYTHON: PASS")


def _groups_flow(
    api: XChatClient,
    core: ChatCore,
    my_id: str,
    partner_id: str,
    both_keys: list[dict],
    signing: list[dict],
) -> None:
    my_keys = [k for k in both_keys if k["user_id"] == my_id]
    signing_pub = core.public_keys()["signing"]

    group_id = api.initialize_group()
    assert group_id.startswith("g"), f"unexpected group id: {group_id!r}"

    # Create with the caller as sole member/admin so the member add below
    # exercises prepare_group_members_change with the partner.
    prep = core.prepare_group_create(my_keys, group_id, [my_id], [my_id])
    body = {
        "conversation_id": group_id,
        "group_members": [my_id],
        "group_admins": [my_id],
        "group_name": "chat-xdk e2e",
        **prep_to_request(prep, signing_pub),
    }
    try:
        api.create_conversation(body)
        members = [my_id]
    except Exception:
        # Some deployments reject single-member groups; fall back to creating
        # with both participants (skipping the member-add below).
        prep = core.prepare_group_create(both_keys, group_id, [my_id, partner_id], [my_id])
        body = {
            "conversation_id": group_id,
            "group_members": [my_id, partner_id],
            "group_admins": [my_id],
            "group_name": "chat-xdk e2e",
            **prep_to_request(prep, signing_pub),
        }
        api.create_conversation(body)
        members = [my_id, partner_id]
    kv = prep["conversation_key_version"]
    key = bytes(prep["conversation_key"])
    print(f"group created: {group_id} with {len(members)} member(s)")

    marker = f"chat-xdk e2e group [python] {int(time.time())}"
    msg = core.encrypt_reply(
        conversation_id=group_id,
        text=marker,
        conversation_key=key,
        conversation_key_version=kv,
    )
    api.send_message(group_id, msg)
    conv_keys = {kv: key}
    one = _await_decrypted(api, core, group_id, conv_keys, signing, msg["message_id"])
    assert message_text(one) == marker and one.get("verified") is True, (
        f"group message round-trip failed: {one}"
    )
    print("group message decrypted + verified")

    if partner_id not in members:
        prep = core.prepare_group_members_change(
            both_keys, group_id, [partner_id], members, [my_id]
        )
        api.add_group_members(
            group_id,
            {"user_ids": [partner_id], **prep_to_request(prep, signing_pub)},
        )
        print(f"group member add: {partner_id} added (key rotated to {prep['conversation_key_version']})")


if __name__ == "__main__":
    main()

File: examples/js/e2e-live.mjs
/**
 * Live end-to-end check against the X Chat API, using the example's real
 * ChatCore (WASM binding) and XChatClient (XDK). Run manually:
 *
 *   CHATXDK_E2E=1 X_ACCESS_TOKEN=... CHAT_PRIVATE_KEYS_B64=... CHAT_SIGNING_KEY_VERSION=... \
 *   CHAT_CONVERSATION_ID=... node e2e-live.mjs
 *
 * Flow (each numbered step asserts against the live API):
 *   1. batch-decrypt inbound history (pagination when a second page exists)
 *   2. rotate the conversation key (prepare -> POST /keys -> decrypt own CKCE)
 *   3. send a threaded reply with an entity + TTL under the rotated key,
 *      fetch it back, decrypt it via the single-event path, and verify it
 *   4. react to the sent message (add + remove), decrypting the add back
 *
 * Optional extras: CHATXDK_E2E_MEDIA=1 also stream-encrypts a media blob,
 * uploads it, sends a message referencing it, then downloads and
 * stream-decrypts it back to the original bytes; CHATXDK_E2E_GROUPS=1 also
 * creates a group (two-signature create), sends a group message, and adds
 * the 1:1 partner as a member.
 *
 * Not part of `node --test` (kept outside test/ and without a .test suffix).
 */
import { ChatCore, messageText, prepToRequest } from "./src/chat-core.mjs";
import { XChatClient } from "./src/x-api.mjs";

const need = (k) => {
  const v = process.env[k];
  if (!v) throw new Error(`missing ${k}`);
  return v;
};
const assert = (cond, msg) => {
  if (!cond) throw new Error(msg);
};
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const token = need("X_ACCESS_TOKEN");
const blob = need("CHAT_PRIVATE_KEYS_B64");
const ver = need("CHAT_SIGNING_KEY_VERSION");
const conv = need("CHAT_CONVERSATION_ID");

function signingFrom(pk, userId) {
  return {
    userId,
    publicKeyVersion: String(pk.publicKeyVersion ?? pk.public_key_version ?? ""),
    publicKey: pk.signingPublicKey ?? pk.signing_public_key ?? "",
    identityPublicKey: pk.publicKey ?? pk.public_key ?? "",
    identityPublicKeySignature:
      pk.identityPublicKeySignature ?? pk.identity_public_key_signature ?? "",
  };
}

/** Public-keys response -> the flat entries the prepare methods take. */
function keyEntries(pks, userId) {
  return pks.map((pk) => ({
    userId,
    publicKey: pk.publicKey ?? pk.public_key ?? "",
    keyVersion: String(pk.publicKeyVersion ?? pk.public_key_version ?? ""),
  }));
}

/**
 * Poll the conversation until the event for `messageId` lands, and return it
 * decrypted via the single-event path (`decryptOne`) as
 * `{ event, b64 }` — the raw base64 alongside, for reply/reaction-by-event.
 *
 * The target envelope is matched by its raw event id before decrypting, so a
 * decrypt failure on our own event (e.g. a broken sign->verify loop) surfaces
 * in the timeout message instead of being silently swallowed.
 */
async function awaitDecrypted(api, core, conversationId, convKeys, signing, messageId, tries = 10) {
  let lastErr;
  for (let i = 0; i < tries; i++) {
    const page = await api.getEvents(conversationId, { maxResults: 25 });
    for (const e of page.data ?? []) {
      const b64 = e.encodedEvent ?? e.encoded_event;
      if (!b64) continue;
      const isTarget = String(e.id ?? "") === messageId;
      let one;
      try {
        one = core.decryptOne(b64, convKeys, signing);
      } catch (err) {
        if (isTarget) lastErr = err;
        continue;
      }
      if (isTarget || String(one.id ?? "") === messageId) {
        // The REST item's id IS the event's sequence id.
        if (!one.sequenceId) one.sequenceId = e.id;
        return { event: one, b64 };
      }
    }
    await sleep(1000);
  }
  throw new Error(
    `event for sent message ${messageId} never appeared` +
      (lastErr ? ` (last decrypt error: ${lastErr})` : ""),
  );
}

const core = await ChatCore.create();
core.loadKeys(blob, ver);
const api = new XChatClient(token);
const myId = await api.getMyUserId();
// The session identity signs everything sent below; per-call senderId /
// signingKeyVersion arguments are gone.
core.setIdentity(myId);

// KeyChange events arrive in meta.conversation_key_events, separate from
// data; they carry the conversation keys and must go into the same
// decryptEvents batch as the data events.
const keyEventsOf = (p) =>
  p.meta?.conversationKeyEvents ?? p.meta?.conversation_key_events ?? [];

// -- 1. Inbound history: batch decrypt (+ pagination when available) --------
const page = await api.getEvents(conv, { maxResults: 10 });
let raw = page.data ?? [];
const keyEventsB64 = [...keyEventsOf(page)];
const nextToken = page.meta?.nextToken ?? page.meta?.next_token;
if (nextToken) {
  const page2 = await api.getEvents(conv, { maxResults: 10, paginationToken: nextToken });
  const raw2 = page2.data ?? [];
  const ids1 = new Set(raw.map((e) => String(e.id)));
  assert(
    raw2.length > 0 && !raw2.some((e) => ids1.has(String(e.id))),
    "pagination made no progress",
  );
  raw = raw.concat(raw2);
  keyEventsB64.push(...keyEventsOf(page2));
  console.log(`pagination: fetched second page with ${raw2.length} events`);
}

const ids = new Set([myId]);
for (const e of raw) {
  const s = String(e.senderId ?? e.sender_id ?? "");
  if (s) ids.add(s);
}
const signing = [];
const pksByUser = new Map();
for (const id of ids) {
  try {
    const pks = await api.getPublicKeys(id);
    pksByUser.set(id, pks);
    for (const pk of pks) signing.push(signingFrom(pk, id));
  } catch {
    /* ignore */
  }
}

const eventsB64 = [
  ...keyEventsB64,
  ...raw.map((e) => e.encodedEvent ?? e.encoded_event).filter(Boolean),
];
let batch = core.decryptBatch(eventsB64, signing);
const decrypted = batch.messages.filter((m) => messageText(m.event)).length;
let convKeys = { ...batch.conversationKeys.keys };
console.log(
  `live inbound messages decrypted: ${decrypted}; conversation keys: ${Object.keys(convKeys).length}`,
);
assert(decrypted > 0, "expected to decrypt at least one live message");

// Canonical conversation_id + partner id + the last inbound message (raw
// b64, for the reply-by-event form below).
let canonicalConv = conv;
let lastInboundB64;
for (const m of batch.messages) {
  const ev = m.event ?? {};
  if (ev.conversationId) canonicalConv = ev.conversationId;
  if (ev.type === "message" && String(ev.senderId) !== myId && ev.sequenceId && m.originalB64) {
    lastInboundB64 = m.originalB64;
  }
}
const partnerId = [...ids].find((id) => id !== myId);
assert(partnerId, "expected a conversation partner among the senders");

// -- 2. Key rotation: prepare -> POST /keys -> decrypt own CKCE -------------
const bothKeys = [
  ...keyEntries(pksByUser.get(myId) ?? [], myId),
  ...keyEntries(pksByUser.get(partnerId) ?? [], partnerId),
];
const prep = core.prepareConversationKeyChange({
  publicKeys: bothKeys,
});
const signingPub = core.publicKeys().signing;
const resp = await api.addConversationKeys(conv, prepToRequest(prep, signingPub));
const data = resp.data ?? {};
assert(
  data.sequenceId ||
    data.sequence_id ||
    data.conversationKeyChangeSequenceId ||
    data.conversation_key_change_sequence_id,
  `key rotation not acknowledged: ${JSON.stringify(resp)}`,
);
const serverConv = data.conversationId ?? data.conversation_id;
console.log(
  `rotated conversation key to version ${prep.conversationKeyVersion}` +
    (serverConv ? `; server conversation_id: ${serverConv}` : ""),
);

// Re-fetch (polling briefly, in case the CKCE has not propagated yet) so our
// own CKCE decrypts and the cache includes the new version.
const kv = prep.conversationKeyVersion;
for (let i = 0; i < 5; i++) {
  const page3 = await api.getEvents(conv, { maxResults: 10 });
  batch = core.decryptBatch(
    [
      ...keyEventsOf(page3),
      ...(page3.data ?? []).map((e) => e.encodedEvent ?? e.encoded_event).filter(Boolean),
    ],
    signing,
  );
  convKeys = { ...batch.conversationKeys.keys };
  if (convKeys[kv]) break;
  await sleep(1500);
}
assert(convKeys[kv], `own rotated CKCE (version ${kv}) did not decrypt+verify`);
const key = convKeys[kv];

// -- 3. Send under the rotated key; fetch back; single-event decrypt --------
const marker = `chat-xdk e2e [js] ${Date.now()}`;
assert(lastInboundB64, "expected a raw inbound event to reply to");
const body = core.encryptReply({
  conversationId: canonicalConv,
  text: `@user ${marker}`,
  conversationKey: key,
  conversationKeyVersion: kv,
  replyToEvent: lastInboundB64,
  entities: [[0, 5, "mention"]],
  ttlMsec: 24 * 60 * 60 * 1000,
});
await api.sendMessage(canonicalConv, body);
console.log(`sent live encrypted message: ${JSON.stringify(marker)}`);

const { event: one, b64: oneB64 } = await awaitDecrypted(
  api,
  core,
  conv,
  convKeys,
  signing,
  body.message_id,
);
assert(messageText(one) === `@user ${marker}`, `round-trip text mismatch: ${JSON.stringify(one)}`);
assert(one.verified === true, "own sent message failed signature verification");
console.log("sent message decrypted + verified via the single-event path");

// -- 4. Reactions: add (round-trip) then remove — both target the raw event --
const add = core.encryptReaction({
  add: true,
  targetEvent: oneB64,
  emoji: "\u{1f44d}",
  conversationKey: key,
  conversationKeyVersion: kv,
});
await api.sendMessage(canonicalConv, add);
const { event: reaction } = await awaitDecrypted(api, core, conv, convKeys, signing, add.message_id);
assert(
  reaction.content?.contentType === "reaction" && reaction.content?.emoji === "\u{1f44d}",
  `expected a reaction event, got ${JSON.stringify(reaction.content)}`,
);
assert(reaction.verified === true, "reaction failed signature verification");
console.log("reaction add decrypted + verified");

const remove = core.encryptReaction({
  add: false,
  targetEvent: oneB64,
  emoji: "\u{1f44d}",
  conversationKey: key,
  conversationKeyVersion: kv,
});
await api.sendMessage(canonicalConv, remove);
console.log("reaction remove sent");

// -- 5. Optional: media — stream-encrypt, upload, send, download, decrypt ----
if (process.env.CHATXDK_E2E_MEDIA === "1") {
  // A deterministic multi-chunk payload, so the incremental encryptor emits
  // several frames and any corruption is byte-attributable.
  const plaintext = Uint8Array.from({ length: 300_000 }, (_, i) => (i * 31 + 7) % 256);
  const ciphertext = core.encryptMedia(plaintext, key);
  const mediaHashKey = await api.uploadMedia(canonicalConv, ciphertext);
  console.log(`encrypted media uploaded: ${mediaHashKey} (${ciphertext.length} bytes)`);

  const mediaMsg = core.encryptReply({
    conversationId: canonicalConv,
    text: `chat-xdk e2e media [js] ${Date.now()}`,
    conversationKey: key,
    conversationKeyVersion: kv,
    attachments: [
      {
        attachment_type: "media",
        media_hash_key: mediaHashKey,
        width: 0,
        height: 0,
        filesize_bytes: plaintext.length,
        filename: "e2e.bin",
        media_type: 5,
      },
    ],
    ttlMsec: 24 * 60 * 60 * 1000,
  });
  await api.sendMessage(canonicalConv, mediaMsg);
  const { event: mediaOne } = await awaitDecrypted(
    api,
    core,
    conv,
    convKeys,
    signing,
    mediaMsg.message_id,
  );
  assert(mediaOne.verified === true, "media message failed signature verification");
  const atts = mediaOne.content?.attachments ?? [];
  const gotKey = atts.find((a) => a.media)?.media?.mediaHashKey;
  assert(gotKey === mediaHashKey, `attachment did not round-trip: ${JSON.stringify(atts)}`);

  const downloaded = await api.downloadMedia(canonicalConv, mediaHashKey);
  const decrypted = core.decryptMedia(downloaded, key);
  assert(
    decrypted.length === plaintext.length && decrypted.every((b, i) => b === plaintext[i]),
    "downloaded media did not decrypt to the original bytes",
  );
  console.log("media downloaded + stream-decrypted to the original bytes");
}

// -- 6. Optional: group create + message + member add ------------------------
if (process.env.CHATXDK_E2E_GROUPS === "1") {
  const myKeys = keyEntries(pksByUser.get(myId) ?? [], myId);
  const groupId = await api.initializeGroup();
  assert(groupId.startsWith("g"), `unexpected group id: ${groupId}`);

  // Create with the caller as sole member/admin so the member add below
  // exercises prepareGroupMembersChange with the partner.
  let gPrep = core.prepareGroupCreate({
    publicKeys: myKeys,
    conversationId: groupId,
    memberIds: [myId],
    adminIds: [myId],
  });
  let members = [myId];
  try {
    await api.createConversation({
      conversationId: groupId,
      groupMembers: members,
      groupAdmins: [myId],
      groupName: "chat-xdk e2e",
      ...prepToRequest(gPrep, signingPub),
    });
  } catch {
    // Some deployments reject single-member groups; fall back to creating
    // with both participants (skipping the member-add below).
    gPrep = core.prepareGroupCreate({
      publicKeys: bothKeys,
      conversationId: groupId,
      memberIds: [myId, partnerId],
      adminIds: [myId],
    });
    members = [myId, partnerId];
    await api.createConversation({
      conversationId: groupId,
      groupMembers: members,
      groupAdmins: [myId],
      groupName: "chat-xdk e2e",
      ...prepToRequest(gPrep, signingPub),
    });
  }
  const gkv = gPrep.conversationKeyVersion;
  const gkey = gPrep.conversationKey;
  console.log(`group created: ${groupId} with ${members.length} member(s)`);

  const gMarker = `chat-xdk e2e group [js] ${Date.now()}`;
  const gBody = core.encryptReply({
    conversationId: groupId,
    text: gMarker,
    conversationKey: gkey,
    conversationKeyVersion: gkv,
  });
  await api.sendMessage(groupId, gBody);
  const { event: gOne } = await awaitDecrypted(
    api,
    core,
    groupId,
    { [gkv]: gkey },
    signing,
    gBody.message_id,
  );
  assert(
    messageText(gOne) === gMarker && gOne.verified === true,
    `group message round-trip failed: ${JSON.stringify(gOne)}`,
  );
  console.log("group message decrypted + verified");

  if (!members.includes(partnerId)) {
    const mPrep = core.prepareGroupMembersChange({
      publicKeys: bothKeys,
      conversationId: groupId,
      newMemberIds: [partnerId],
      currentMemberIds: members,
      currentAdminIds: [myId],
    });
    await api.addGroupMembers(groupId, {
      userIds: [partnerId],
      ...prepToRequest(mPrep, signingPub),
    });
    console.log(
      `group member add: ${partnerId} added (key rotated to ${mPrep.conversationKeyVersion})`,
    );
  }
}

console.log("E2E JS: PASS");
